Added recovery code option to 2fa

alitoshmatov opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
User can use recovery code instead of two-factor authentication code by switching to recovery code input
### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/22335
PROPOSAL: https://github.com/Expensify/App/issues/22335#issuecomment-1628319424


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Have a user whose 2fa is enabled
2. Sign in with this user
3. At two-factor authentication step make sure `Use recovery code` text link is present
4. Make sure pressing it switches input type to simple text input 
5. Enter recovery code
6. Make sure recovery code is being validated correctly and you can sign in with correct recovery code
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same as tests
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
7. Upload an image via copy paste
8. Verify a modal appears displaying a preview of that image
--->
Same as tests
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/59907218/d65c8a15-655e-445f-812e-26107ed9daae


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/59907218/2a6ddf40-777b-4fd4-a8ef-5808ae06bd73


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/59907218/3cb94da5-6021-46ad-a6b3-fab7bcf16c07


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/59907218/53f21571-0454-45a5-a82f-4a63762a95f4


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/59907218/4fe182fd-2fa9-4a85-83ed-ef46e228c264


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/59907218/26e49726-9323-4cfe-894e-beb3ad4b082a


<!-- add screenshots or videos here -->

</details>

alitoshmatov committed:
Added recovery code feature

alitoshmatov committed:
Added focus delay so that keyboard behaves smoothlier

alitoshmatov commented on src/languages/en.js:
```diff
@@ -567,6 +567,15 @@ export default {
         copy: 'Copy',
         disable: 'Disable',
     },
+    recoveryCodeForm: {
+        error: {
+            pleaseFillRecoveryCode: 'Please enter your recovery code',
+            incorrectRecoveryCode: 'Incorrect recovery code. Please try again.',
+        },
+        useRecoveryCode: 'Use recovery code',
+        recoveryCode: 'Recovery code',
+        use2fa: 'Use two-factor authentication code',
+    },
```
Made this texts similar to 2fa error texts. But still need confirmation for this and spanish version

alitoshmatov committed:
Merge branch 'main' of https://github.com/Expensify/App into 22335/2fa-recovery-code

alitoshmatov commented:
@abdulrahuman5196 Friendly bump

abdulrahuman5196 commented on src/languages/en.js:
```diff
@@ -567,6 +567,15 @@ export default {
         disable: 'Disable',
     },
+    recoveryCodeForm: {
+        error: {
```
@MonilBhavsar We need to have Copy verification for all these strings?

abdulrahuman5196 replied:
Can we kindly add the label for text verification

alitoshmatov replied:
@MonilBhavsar Friendly bump

MonilBhavsar replied:
Sorry, I missed this. Commented on the issue. Once verified, we can get a spanish translation

alitoshmatov replied:
Translations are confirmed - https://github.com/Expensify/App/issues/22335#issuecomment-1684060434

alitoshmatov committed:
Merge branch 'main' of https://github.com/Expensify/App into 22335/2fa-recovery-code

alitoshmatov commented:
Translations are confirmed, and updated code with the latest main branch and solved conflicts.

Sorry for force push, I have changed my system and had some problems authenticating with github. 

It is ready for review now
@abdulrahuman5196

abdulrahuman5196 commented:
@alitoshmatov We should clear recovery code data as well here https://github.com/Expensify/App/pull/23390/files#diff-873bd2e06c31caa8e7ca519b38cbd825dad372a68ea5f4e4f00bb27f60b9a9c7R185

abdulrahuman5196 commented:
@MonilBhavsar The recovery code field doesn't have auto-check after entering recovery code. Like we have to press signin after the recovery code has entered in the text input. But both magic code and 2FA code auto hit API after just entering value.
Do we want that? If so in this issue or followup issue?

abdulrahuman5196 commented:
@alitoshmatov There are multiple useEffects involving clearing/focusing of input based on magicCode and 2FA code status, can we check and include recovery code as well to logics or create useEffects for recovery code in similar way

abdulrahuman5196 commented on src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js:
```diff
@@ -183,6 +196,22 @@ function BaseValidateCodeForm(props) {
+        setRecoveryCode('');
+        setTwoFactorAuthCode('');
+
+        setFormError((prevError) => ({...prevError, recoveryCode: '', twoFactorAuthCode: ''}));
```
Why do we need prevError?

alitoshmatov committed:
Added clearing recovery code to clearLocalSignInData

alitoshmatov commented:
> @alitoshmatov We should clear recovery code data as well here https://github.com/Expensify/App/pull/23390/files#diff-873bd2e06c31caa8e7ca519b38cbd825dad372a68ea5f4e4f00bb27f60b9a9c7R185

Added clearing recovery code there.


> @alitoshmatov There are multiple useEffects involving clearing/focusing of input based on magicCode and 2FA code status, can we check and include recovery code as well to logics or create useEffects for recovery code in similar way

I don't think there is a need for handling recovery code in similar cases

@abdulrahuman5196

abdulrahuman5196 commented:
>There are multiple useEffects involving clearing/focusing of input based on magicCode and 2FA code status, can we check and include recovery code as well to logics or create useEffects for recovery code in similar way

>I don't think there is a need for handling recovery code in similar cases.

@alitoshmatov  Could you kindly add information on why? Anyways i wasn't able to find any visible issues based on the same, but we need to be sure if we don't want to add similar effects. The reason I think is if something is applicable to 2FA code then the same logic should be applicable for recovery codes as well.

MonilBhavsar commented:
> @MonilBhavsar The recovery code field doesn't have auto-check after entering recovery code. Like we have to press signin after the recovery code has entered in the text input. But both magic code and 2FA code auto hit API after just entering value. Do we want that? If so in this issue or followup issue?

By auto check, you mean auto submit? I think it is only for those numeric inputs. We can leave it here

alitoshmatov commented:
Not sure why this exists, but it looks like it was introduced when refactoring to functional components, and these changes were commited under `Guarantee clearing after state change` message. As this message suggests, these changes make sure inputs are cleared when user goes from general validation code to 2fa code inputs.

https://github.com/Expensify/App/blob/b43c20f4336ea941a33fb52ce5d6ec0195956af5/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js#L119-L131

This is one is here since we store `validateCode` in onyx state, and it makes sure they are matched.
https://github.com/Expensify/App/blob/b43c20f4336ea941a33fb52ce5d6ec0195956af5/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js#L105-L110

And these are for managing focus state of validate code input, not sure why we need this since providing autofocus should work properly, and it seems redundant.
https://github.com/Expensify/App/blob/b43c20f4336ea941a33fb52ce5d6ec0195956af5/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js#L91-L103

To conclude, most of these `useEffects` are here to manage `MagicCodeInput`. But I think there are some redundant code and this could be optimized. Anyways I think there is no need for similar changes for simple `TextInput`

@abdulrahuman5196 Let me know what you think.

alitoshmatov committed:
Merge branch 'main' of https://github.com/Expensify/App into 22335/2fa-recovery-code

alitoshmatov commented:
@abdulrahuman5196 Friendly bump!

MonilBhavsar commented:
@abdulrahuman5196 👋

MonilBhavsar commented:
@abdulrahuman5196 could you please take a look

abdulrahuman5196 commented:
Hi, yes will take a look into it today. I was focusing more on proposals since we had PR merge freeze.

alitoshmatov committed:
Merge branch 'main' of https://github.com/Expensify/App into 22335/2fa-recovery-code

alitoshmatov commented:
Resolved merge conflicts

abdulrahuman5196 commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->




</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->




</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->





</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->





</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->




</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->




</details>

abdulrahuman5196 reviewed (approved):
Changes looks good and works well. Reviewers checklist is also complete.

All yours. @MonilBhavsar 

🎀 👀 🎀
C+ Reviewed

abdulrahuman5196 commented:
Note: I see a slight non-smooth transition on mobile safari alone when switching between inputs. But I don't think its related to this PR since it could be related to re-render on switching input and focus is already not better in safari. 

cc: @MonilBhavsar

MonilBhavsar commented:
We also need to change the text above the input that says - `Please enter your authenticator code` to `Please enter your recovery code`

alitoshmatov commented:
@abdulrahuman5196 
The message above inputs is managed in SinInPage
 https://github.com/Expensify/App/blob/5d5b956dfba80629f17e95a3eeb753dc37ca5580/src/pages/signin/SignInPage.js#L127-L130

I am thinking to shift `isUsingRecoveryCode` state on level up from `BaseValidateCodeForm` to `SingInPage` this way we can keep track of the state and change the message accordingly.

What do you think about this?

abdulrahuman5196 commented:
>I am thinking to shift `isUsingRecoveryCode` state on level up from `BaseValidateCodeForm` to `SingInPage` this way we can keep track of the state and change the message accordingly.


I don't understand this. How do to plan to move a internal state value to its parent component? Currently SignInPage determines the text above input from onyx.

alitoshmatov commented:
@abdulrahuman5196 I mean, we declare `isUsingRecoveryCode` state in `SignInPage` and pass its value, and change function to `BaseValidateCodeForm`, thus making this state available in both places. We add additional condition using this state to determine the message.

Another way is to declare new state in onyx data and keep track of if user is using recovery code

MonilBhavsar commented:
To be honest, a new Onyx data for maintaining state or user method sounds overkill. Would be great if we can do with `state`

alitoshmatov committed:
Shifted  state to signInPage since it was needed to determine welcome text

alitoshmatov committed:
Added eslint rule exception

alitoshmatov commented:
@abdulrahuman5196 Moved state to `SignInPage` from `BaseValidateCodeForm`, thus making welcome text respond when switched to recovery code

abdulrahuman5196 commented:
> @abdulrahuman5196 Moved state to `SignInPage` from `BaseValidateCodeForm`, thus making welcome text respond when switched to recovery code

@MonilBhavsar I am not in that much agreement to this approach, but still i agree on the point of onyx being overkill as well. So I am 50-50 on this. Could you kindly give a quick look at the new changes and let me know if this kind of state management is fine?

MonilBhavsar commented:
That looks good to me. We're taking it one component up to pass the correct `welcomeText` prop, right? @abdulrahuman5196 anything specific you think does not follow good state management practice?

alitoshmatov commented:
@abdulrahuman5196 Friendly bump

abdulrahuman5196 commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->




</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->




</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->





</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->





</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->




</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->




</details>


https://github.com/Expensify/App/assets/46707890/7ab63a4a-da57-4486-9951-fc0478dd2429


https://github.com/Expensify/App/assets/46707890/53301d7f-0476-497b-ab71-22eca18fc490


https://github.com/Expensify/App/assets/46707890/d4b2d55e-d94d-41c2-a0b2-1259e42e0595


https://github.com/Expensify/App/assets/46707890/1a2a3c1b-b78a-4d80-b0a9-92d009d4f820


https://github.com/Expensify/App/assets/46707890/adccc782-cf80-47bc-8d2f-65dce9f4bcdd

abdulrahuman5196 reviewed (approved):
Changes looks good and works well. Reviewers checklist is also complete.

All yours. @MonilBhavsar 

🎀 👀 🎀
C+ Reviewed

MonilBhavsar commented:
@alitoshmatov could you please merge main in this branch

alitoshmatov committed:
Merge branch 'main' of https://github.com/Expensify/App into 22335/2fa-recovery-code

alitoshmatov commented:
@MonilBhavsar Done

MonilBhavsar reviewed (approved):
Looks good and works well! Thank you!!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6274759566) to staging by https://github.com/MonilBhavsar in version: 1.3.73-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6297348186) to production by https://github.com/mountiny in version: 1.3.73-1 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6297355879) to staging by https://github.com/MonilBhavsar in version: 1.3.74-0 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6342870850) to production by https://github.com/chiragsalian in version: 1.3.74-3 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

getusha commented on src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.js:
```diff
@@ -226,33 +279,61 @@ function BaseValidateCodeForm(props) {
+                            accessibilityLabel={props.translate('recoveryCodeForm.recoveryCode')}
+                            value={recoveryCode}
+                            onChangeText={(text) => onTextInput(text, 'recoveryCode')}
+                            maxLength={CONST.RECOVERY_CODE_LENGTH}
```
This caused a regression here https://github.com/Expensify/App/issues/35947
When we paste a recovery code with spaces at the beginning it will cut out some letters from the end.
